### ISSUE TITLE
Change `noDefault` to default to `true`?

### ISSUE THREAD

**Issue by KATT** [Floby/node-envie#8]
Personally I'd prefer `noDefaults` to be `true` as default since we're dealing with sensitive things and I don't ever want to leak anything that hasn't been explicitly defined.

Sorry for all the issues. I've been looking around for nice env var handling and your package is really sweet and close to what I'd ideally want!

**Comment by Floby** [comment]
Thanks a lot for your interest, I'm glad you found this before reaching the same conclusion as I did : code my own :D

I understand your worry, considering that some people define default values as their `production` environment.
The defaults are really useful from my point of view. I like to consider that default values for confiugration should be the one used in the most frequent deployment target. I consider the developer's machine to be the most frequent target so I usually define defaults as "what works when you run `npm start`". These are almost never sensitive information.

Therefore, I'd rather keep the default value as the _default_ behaviour but I'm happy to add a disclaimer regarding what to put in those in the README.

**Comment by KATT** [comment]
I started coding my own but then I found yours, haha 👯‍♂️ 

So, the problem I have is that I can access things from `process.env` that I haven't defined in my joi schema. That it's called `noDefaults` is a bit misleading.

## Now, example:

```js
const { PORT, NODE_ENV } = Envie({
  PORT: Joi
    .number()
    .min(0)
    .default(0)
    .description('Port on which the HTTP server will listen'),
}).values()

console.log(PORT); // 0 or process.env.PORT (I'm fine with this)
console.log(NODE_ENV); // 'development', etc. 
```

Not fine with that you can access `NODE_ENV` in this example that we have not defined it in the our schema. It might lead to issues with false positive type issues. I'd prefer it to be strict.

I'd even want to have `.values()` to implement a `Proxy` that throws an `Error` if you try to access undefined keys.

## Ideal

```js
const env = Envie({
  PORT: Joi
    .number()
    .min(0)
    .default(0)
    .description('Port on which the HTTP server will listen'),
}).values() 

const NODE_ENV = env.NODE_ENV; // throws new Error('Key `NODE_ENV` is not defined in your schema');
```

**Comment by KATT** [comment]
Maybe I got this confused... hold on, playing around and will do a PR

**Comment by Floby** [comment]
I actually wrote that in the [article](https://blog.octo.com/applications-node-js-a-12-facteurs-partie-1-une-base-de-code-saine/) which announced this module (untranslated unfortunately) But maybe Google can translate not too badly :D

**Comment by Floby** [comment]
Oh I see what you mean. This behaviour is closest to what `allowUnknown` does in Joi.
I think I allow it mostly because I thought it would hurt adoption to throw on otherwise defined values (although not _described_).
I think we could add an option to throw when trying to get undescribed keys. Would that be good for you?

**Comment by Floby** [comment]
Throwing on `object.key` might be a little too unexpected :/

**Comment by KATT** [comment]
Exactly, this issue is about not passing that `allowUnknown`, I don't like unexpected side-effects, if I define a schema I want it to be strict :)

> Throwing on `object.key` might be a little too unexpected :/

Maybe. But it could be a proxy just when using the `values()` one. I've used something similar in [awilix's `cradle`](https://github.com/jeffijoe/awilix#containercradle) that works really nicely. If I try to _resolve_ a module through the cradle (by `cradle.myModule`) and `myModule` isn't explicitly defined it throws an Error. That sort of behaviour almost makes JS feel like real programming and you get more confidence in things. 

Let's say for instance this `values()`-result is something I pass around in my app and I refactor to remove an env from my schema 0 as there's no way of tracing my usage of `values` easily it's nice to actually throw an error rather than having a variable somewhere being `undefined` which might lead to some bugs.

**Comment by KATT** [comment]
Anyway, the proxy stuff is a separate discussion, let's park that for now! Sorry to side-track 

In my opinion the following changes should be made

1. Rename `noDefaults` to `allowUnknown` to be consistent
1. Change the default value to `true`

**Comment by Floby** [comment]
1. No these are the exact parameters sent to Joi and meant to reflect their behaviour. I made the choice of the adherence to Joi and would like to keep the semantic. `noDefaults` means not to return the default value when the specified value is undefined. `allowUnknown` means that if the key is unknown (not described in the schema) it should not consider it invalid. `stripUnknown` means undescribed keys should be removed from the result. Considering this, noDefaults is the only one which translate quite perfectly to `envie`, the other ones no quite so.
2. I think the point is irrelevant because I don't think we are talking about the same thing.

There is a difference between _undefined_ keys and _undescribed_ keys.

At the moment when we `.get(key)` with default options:

1.  if described(key) && defined(value): →validate and return value, throw if invalid
2.  if !described(key) && defined(value): →return value, no throw 
3.  if !described(key) && !defined(value)): → return undefined, no throw
4.  if described(key) && !defined(value): → return default value or throw

With the `noDefaults` on:

1.  if described(key) && defined(value): →validate and return value, throw if invalid
2.  if !described(key) && defined(value): →return value, no throw 
3.  if !described(key) && !defined(value)): → return undefined, no throw
4.  if described(key) && !defined(value): → return default value if set or return undefined or throw

I understand that you would like:

1.  if described(key) && defined(value): →validate and return value, throw if invalid
2.  if !described(key) && defined(value): →throw
3.  if !described(key) && !defined(value)): → throw
4.  if described(key) && !defined(value): → return default value if set or return undefined or throw

Does that make sense ?
